### ISSUE TITLE
SWORD: stop calling DeleteDataFileCommand

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2222]
@scolapasta indicated that the SWORD code needs to be modified to not call DeleteDataFileCommand. It should instead call UpdateDatasetCommand.

**Comment by pdurbin** [comment]
I made the change in b62334a which @scolapasta might want to review.

Passing to QA. I only tested deleting files from **unpublished** datasets, which seems to work fine.

**Comment by kcondon** [comment]
Tested, works on an unpublished draft but throws exception on a published dataset and on a draft of a published dataset. V4.0 failed with an error explaining cannot delete file of a published dataset and worked on a draft of a published dataset.

Stack trace:

[2015-06-05T14:50:25.779-0400] [glassfish 4.1] [WARNING] [] [javax.enterprise.web] [tid: _ThreadID=50 _ThreadName=jk-connector(4)] [timeMillis: 1433530225779] [levelValue: 900] [[
  StandardWrapperValve[edu.harvard.iq.dataverse.api.datadeposit.SWORDv2MediaResourceServlet]: Servlet.service() for servlet edu.harvard.iq.dataverse.api.datadeposit.SWORDv2MediaResourceServlet threw exception
javax.ejb.EJBException
        at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:748)
        at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:698)
        at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:503)
        at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4566)
        at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2074)
        at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2044)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
        at com.sun.proxy.$Proxy752.submit(Unknown Source)
        at edu.harvard.iq.dataverse.__EJB31_Generated__EjbDataverseEngine__Intf____Bean__.submit(Unknown Source)
        at edu.harvard.iq.dataverse.api.datadeposit.MediaResourceManagerImpl.deleteMediaResource(MediaResourceManagerImpl.java:157)
        at org.swordapp.server.MediaResourceAPI.delete(MediaResourceAPI.java:356)
        at edu.harvard.iq.dataverse.api.datadeposit.SWORDv2MediaResourceServlet.doDelete(SWORDv2MediaResourceServlet.java:54)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:713)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
        at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:295)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
        at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:205)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:415)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:201)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:175)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:561)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:565)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:545)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Cannot merge an entity that has been removed: edu.harvard.iq.dvn.core.study.FileMetadata[id=224]
        at org.eclipse.persistence.internal.sessions.MergeManager.registerObjectForMergeCloneIntoWorkingCopy(MergeManager.java:1063)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChangesOfCloneIntoWorkingCopy(MergeManager.java:557)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChanges(MergeManager.java:313)
        at org.eclipse.persistence.mappings.CollectionMapping.mergeIntoObject(CollectionMapping.java:1645)
        at org.eclipse.persistence.internal.descriptors.ObjectBuilder.mergeIntoObject(ObjectBuilder.java:4132)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChangesOfCloneIntoWorkingCopy(MergeManager.java:594)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChanges(MergeManager.java:313)
        at org.eclipse.persistence.mappings.CollectionMapping.mergeIntoObject(CollectionMapping.java:1645)
        at org.eclipse.persistence.internal.descriptors.ObjectBuilder.mergeIntoObject(ObjectBuilder.java:4132)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChangesOfCloneIntoWorkingCopy(MergeManager.java:594)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChanges(MergeManager.java:313)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.mergeCloneWithReferences(UnitOfWorkImpl.java:3521)
        at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.mergeCloneWithReferences(RepeatableWriteUnitOfWork.java:384)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.mergeCloneWithReferences(UnitOfWorkImpl.java:3481)
        at org.eclipse.persistence.internal.jpa.EntityManagerImpl.mergeInternal(EntityManagerImpl.java:542)
        at org.eclipse.persistence.internal.jpa.EntityManagerImpl.merge(EntityManagerImpl.java:519)
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper.merge(EntityManagerWrapper.java:305)
        at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetCommand.save(UpdateDatasetCommand.java:149)
        at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetCommand.execute(UpdateDatasetCommand.java:75)
        at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetCommand.execute(UpdateDatasetCommand.java:37)
        at edu.harvard.iq.dataverse.EjbDataverseEngine.submit(EjbDataverseEngine.java:178)
        at sun.reflect.GeneratedMethodAccessor3729.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
        at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
        at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
        at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:46)
        at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
        at sun.reflect.GeneratedMethodAccessor2474.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
        at sun.reflect.GeneratedMethodAccessor3150.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
        at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
        at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
        at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
        ... 44 more
]]

**Comment by kcondon** [comment]
@scolapasta Not sure why this was assigned to QA: @pdurbin was this ever looked at, in your opinion should it be working now?

**Comment by pdurbin** [comment]
@kcondon either you or I should probably check to make sure we aren't getting that exception. I do have a REST Assured set for deleting a file now: 3e9ca28

**Comment by scolapasta** [comment]
@pdurbin can you check this issue?

**Comment by pdurbin** [comment]
A better title for this issue would be "Follow up on Unintuitive behaviors when deleting files via SWORD API in #1784" because that's really what this issue is about.

Basically, back in April/May 2015 or so @scolapasta worked on #1784 and then he and I created this issue for me to change the SWORD code to use the code that has his fix in it. In b62334a (part of 4.0.1) I made this change, and since then SWORD has been calling `UpdateDatasetCommand(dataset, user, filesToDelete)` rather than `DeleteDataFileCommand(fileToDelete, user)`. Then I passed it to QA.

This issue indicates that @kcondon did some testing in June 2015 and noted problems (exceptions thrown) for the following cases:

- deleting a file from a published dataset
- deleting a file from a post-publication draft

This is where I focused my efforts today, writing some Rest Assured tests in 9153e18. The new assertions are in a test called `testDeleteFiles` which can be seen at https://github.com/IQSS/dataverse/blob/9153e18a762017841daf4dd531dd8ac9d1e4cb2d/src/test/java/edu/harvard/iq/dataverse/api/SwordIT.java#L230

Basically, I upload three files to a new dataset, delete one of them, publish the dataset, delete another file (causing the dataset to go into draft mode), and delete the last file. Here's a screenshot:

![screen shot 2016-02-02 at 1 22 38 pm](https://cloud.githubusercontent.com/assets/21006/12760199/3e89e7e6-c9b3-11e5-8164-3b77416ac60e.png)

I think these tests are nice (and I fixed some other problems with the test suite) so I made a pull request at #2906 to get these merged into develop.

I'm going to move this ticket into QA. To test, I would recommend looking at the original bug report by @rliebz in #1784 about "unintuitive behaviors when deleting files via SWORD API."

**Comment by pdurbin** [comment]
Testing should also include the scenario here: SWORD: deleting a file from a published version (not a draft) creates a draft but doesn't delete the file · Issue #2464 

However, I just re-tested that bug and it's still a bug. Obviously, the Rest Assured tests I wrote so far for this issue haven't caught it. I'm pulling this out of QA.

@scolapasta please advise. You're the one who worked on the fixes I mentioned in https://github.com/IQSS/dataverse/issues/2222#issuecomment-178752740 . Any ideas?

**Comment by pdurbin** [comment]
I just improved the Rest Assured tests (still in a branch) in 401d840 so that #2464 is now being exercised. @scolapasta I'm hoping you can take the first look since you worked on the last fix. I'll put this in 4.3 and assign it to you to make sure we talk about it.

**Comment by pdurbin** [comment]
> I just improved the Rest Assured tests (still in a branch) in 401d840

We just merged these tests into the develop branch.

**Comment by pdurbin** [comment]
#2464 is still open which is enough of a reminder that something is a little wonky with deleting files via API. I'm hoping we'll learn more when we work on that issue or #3419.
